### ISSUE TITLE
matching prediction logits to ground truth labels

### ISSUE THREAD

**Issue by omidgh1379** [ultralytics/ultralytics#20927]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello 

Im  trying to apply tempreture scaling for yolov8 and then plot the reliablity diagram. I managed to change the source code to get the logits from yolo and then I wrote a function to match the logits with the ground truth. 

First I applied Nms to the detections to filter out duplicate and then, I tried to get the max iou for each gound truth and if the max iou is higher than a threshold, I  considered as true postive. then, add the label and logits together that Im later gonig to use to plot realiblity. 

I want to ask is this logits correct ? and where can I find similar function that you applied for matching. 

thanks 

def _match_targets_to_outputs(self, outputs, targets):  # for each image 
        
        labels = []
        logits  = []
    

        logits_for_nms = outputs.clone()  # Clone to avoid modifying original logits
        logits_for_nms = logits_for_nms.unsqueeze(0).transpose(-1, -2)  # Add batch dimension and transpose to match expected shape
        confidence = torch.sigmoid(logits_for_nms[: , 4:14 , :]  )
        bbox = logits_for_nms[: , 0:4 , :]   # Extract bounding box coordinates (x, y, x, y)
        combined = torch.cat((bbox , confidence) , dim = 1)  # Remove batch dimension if present

        nms_detections = non_max_suppression(combined, conf_thres=0.5,nc = 10, iou_thres=0.45, classes=None, agnostic=False, max_det=300 , in_place=False , return_idxs=True)

      

        original_indices = nms_detections[1][0] # Get the original indices of detections before NMS
        

        if len(nms_detections[0][0]) == 0:
            return None, None
            
        pred_bboxes_xyxy = nms_detections[0][0][..., :4]  # Extract bounding boxes (x, y,x, y)
        targets_bboxes = targets['bboxes']   #[N , 4]

        targets_bboxes_xyxy = torch.cat([targets_bboxes[: , :2]- targets_bboxes[: , 2:4]/2 , targets_bboxes[: , :2]+ targets_bboxes[: , 2:4]/2], dim = 1)
        
        image_height, image_width = 640, 640  # Replace with your image size
        # Normalize bounding boxes to [0, 1] range
        pred_bboxes_xyxy[:, 0] /= image_width   # Normalize x
        pred_bboxes_xyxy[:, 1] /= image_height  # Normalize y
        
        pred_bboxes_xyxy[:, 2] /= image_width   # Normalize x2
        pred_bboxes_xyxy[:, 3] /= image_height  # Normalize y2
        
      
        

        iou_matrix = box_iou(targets_bboxes_xyxy, pred_bboxes_xyxy)   # N , M
        
       
        
        max_iou, max_iou_index = torch.max(iou_matrix , dim = 1)   # shape (N , )  # which detection index in nms has the maximum iou with each target
        
        
        
        for i in range(len(max_iou_index)):
            if iou_matrix[i , max_iou_index[i]] > 0.45:
               
                labels.append(targets['cls'][i])
                
                 # Use the original index to get logits from the original outputs
                original_detection_idx = original_indices[max_iou_index[i]]
                logits.append(outputs[original_detection_idx , 4:14])

            

        print(f"Number of matched logits: {len(logits)}")
        print(f"Number of matched labels: {len(labels)}")
        if len(logits) == 0:
            
            return None, None 
        

        logits= torch.stack(logits , dim = 0) if len(logits)>0 else None
        labels= torch.stack(labels , dim = 0) if len(labels)>0 else None        
        
            
        return logits  , labels

**Comment by UltralyticsAssistant** [comment]
👋 Hello @omidgh1379, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will assist you further soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for comprehensive guides, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, which may address your question about matching predictions to ground truth and working with logits.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much relevant information as possible, including dataset image samples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, join [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- For sharing and exploring threads, check out our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

To ensure you are not encountering issues already resolved in the latest release, upgrade to the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following up-to-date, verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/), preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

🎯 Thank you for your detailed code and explanation! An Ultralytics engineer will review your approach and provide further guidance on matching functions and best practices for working with logits.

**Comment by glenn-jocher** [comment]
Your approach looks reasonable for matching predictions to ground truth! The Ultralytics codebase has a similar function called [`match_predictions`](https://docs.ultralytics.com/reference/engine/validator/#match_predictions) in `ultralytics/engine/validator.py` that you might find helpful as a reference.

The `match_predictions` function takes predicted classes, true classes, and an IoU matrix, then returns a correct tensor indicating matches across different IoU thresholds. It handles the matching logic similar to what you're implementing.

You can also look at [`process_batch`](https://docs.ultralytics.com/reference/utils/metrics/#process_batch) in `ultralytics/utils/metrics.py` which does detection matching for confusion matrix updates, and the target assignment logic in `ultralytics/utils/tal.py` with functions like [`get_box_metrics`](https://docs.ultralytics.com/reference/utils/tal/#get_box_metrics).

Your logic of using max IoU per ground truth with a threshold is sound for temperature scaling purposes, though you might want to consider handling the case where multiple predictions match the same ground truth (currently your approach could assign the same detection to multiple ground truths).
